graph/graph.go

```diff
@@ -27,11 +28,64 @@ import (
 	"github.com/docker/docker/runconfig"
 )
 
+type retainedLayers struct {
+	layers map[string][]string
+	sync.Mutex
+}
+
+func (r *retainedLayers) Add(layerIDs []string, imgID string) {
+	r.Lock()
+	defer r.Unlock()
+	for _, layerID := range layerIDs {
+		if imgs, exists := r.layers[layerID]; exists {
+			r.layers[layerID] = append(imgs, imgID)
+		} else {
+			r.layers[layerID] = []string{imgID}
+		}
+	}
+}
+
+func (r *retainedLayers) Delete(layerIDs []string, imgID string) {
```

Comment:
I'm not sure if it was discussed already, but this delete logic could probably be made more efficient if we use

```go
map[string]map[string]struct{}
```

for storing the strings in a set rather than a slice.

The method could then be:

```go
func (r *retainedLayers) Delete(layerIDs []string, imgID string) {
	r.Lock()
	defer r.Unlock()

	for _, layerID := range layerIDs {
		imgIDs, exists := r.layers[layerID]
		if !exists {
			continue // No references for this layerID.
		}

		delete(imgIDs, imgID)

		if len(imgIDs) == 0 {
			delete(r.layers, layerID) // Delete any empty reference set.
		}
	}
}
```